File: 98. Validate Binary Search Tree.js
/*
98. Validate Binary Search Tree
Medium

Given a binary tree, determine if it is a valid binary search tree (BST).
Assume a BST is defined as follows:
The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.

Example 1:
    2
   / \
  1   3
Input: [2,1,3]
Output: true
Example 2:
    5
   / \
  1   4
     / \
    3   6
Input: [5,1,4,null,null,3,6]
Output: false
Explanation: The root node's value is 5 but its right child's value is 4.
*/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
  const helper = (root, max = Infinity, min = -Infinity) => {
    if (!root) return true;
    if (root.val >= max || root.val <= min) return false;
    if (root.left) {
      if (root.val <= root.left.val) return false;
    }
    if (root.right) {
      if (root.val >= root.right.val) return false;
    }
    return (
      helper(root.left, root.val, min) && helper(root.right, max, root.val)
    );
  };
  return helper(root);
};

File: python/98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if root == None:
            return True
        cur = root
        positive_infnity = float('inf') 
        negative_infnity = float('-inf')
        self.flag = True
        def helper(node, min, max):
            if node.val <= min or node.val >= max:
                self.flag = False
            if node.left:
                helper(node.left, min, node.val)
            if node.right:
                helper(node.right, node.val, max)
        helper(cur, negative_infnity, positive_infnity)
        return self.flag